### Imports and loads

In [23]:
from typing import List

import pickle as pkl
import numpy as np
import html
from pathlib import Path

from fastai.text import *
from sklearn.model_selection import train_test_split

In [2]:
DATA_PATH = Path('DATA/')

### Dataset properties, inspection, tokenization

In [3]:
DATASET_NAME = 'x_and_y_cleaned.pkl'
with open(DATA_PATH/DATASET_NAME, 'rb') as f:
    articles, categories = pkl.load(f)

In [4]:
# Label None as 'none'
categories = ['none' if not x else x for x in categories]

In [5]:
CLASSES = sorted(list(set(categories)))
ARTICLE_COUNT = len(articles)
CLASS_COUNT = len(CLASSES)
BOS = 'xbos'  # beginning-of-sentence tag
FLD = 'xfld'  # data field tag
MAX_SIZE = 250

MAX_VOCAB = 60000
min_freq = 5

print(ARTICLE_COUNT)
print(CLASS_COUNT)

48514
138


#### inspect

In [6]:
# Class balance check:
freq = Counter(o for o in categories)
freq.most_common(25)

[('uudised/eesti', 16451),
 ('melu/elu', 5883),
 ('uudised/maailm', 4285),
 ('uudised/krimi', 2211),
 ('televeeb/tvuudised', 1462),
 ('arvamus/kommentaar', 1342),
 ('naine/naised', 1163),
 ('naine/suhted', 1155),
 ('melu/seltskond', 969),
 ('sport/jalgpall', 885),
 ('naine/ilu', 817),
 ('uudised/kiiksud', 544),
 ('sport/korvpall', 541),
 ('tervis/keha', 502),
 ('blogid/londonilustiblogi', 480),
 ('uudised/ilm', 461),
 ('arvamus/juhtkiri', 454),
 ('sport/varia', 413),
 ('raha/kodu', 353),
 ('meedia/galeriid', 334),
 ('arvamus/repliik', 324),
 ('melu/saund', 323),
 ('melu/sunagukolab', 322),
 ('sport/kergejoustik', 316),
 ('meedia/videod', 315)]

In [7]:
print(CLASSES)

['aiaeri', 'arvamus', 'arvamus/intervjuu', 'arvamus/juhtkiri', 'arvamus/karikatuur', 'arvamus/kommentaar', 'arvamus/koomiks', 'arvamus/lugejakiri', 'arvamus/nadalatipud', 'arvamus/repliik', 'arvamus/seisukoht', 'blogid/avastaeestimaad', 'blogid/aveameerikas', 'blogid/filmiblogi', 'blogid/hollandiblogi', 'blogid/indoneesiablogi', 'blogid/jumestusblogi', 'blogid/korvpallimm', 'blogid/lehesaba', 'blogid/londonilustiblogi', 'blogid/malluka', 'blogid/meistriteblogi', 'blogid/moeajakiri', 'blogid/moekeeris', 'blogid/motteid', 'blogid/muusikablogi', 'blogid/opetajablogi', 'blogid/psyhholoogiablogi', 'blogid/pulmablogi', 'blogid/raamatublogi', 'blogid/raha', 'blogid/seljakotigablogi', 'blogid/spordiblogi', 'blogid/teleblogi', 'blogid/trenniblogi', 'blogid/valdojahilo', 'blogid/yksikvanem', 'eestinaine/elud-inimesed', 'eriline/horoskoop', 'eriline/mystika', 'joulud', 'kroonika/eesti', 'lemmikloom', 'linnaleht/arvamus', 'linnaleht/dilaila', 'linnaleht/karikatuur', 'linnaleht/kodusedlood', 'linna

In [8]:
# Dataset examples:
index = 0
print('ARTICLE: ', articles[index][0:110], '...')
print('CATEGORY: ', categories[index])

ARTICLE:  Kas parima aastavahetuse programmi pani eetrisse ETV, Kanal 2 või hoopis TV3? ETVst näegid vaatajad saateid "V ...
CATEGORY:  televeeb/tvuudised


In [54]:
# Get median/average word count
print(np.median([len(x.split(' ')) for x in articles]))
print(np.mean([len(x.split(' ')) for x in articles]))

261.0
387.0457805994146


In [6]:
# One hot encoding
# labels = []
# for x in categories:
#     y = [0 for x in range(CLASS_COUNT)]
#     y[CLASSES.index(x)] = 1
#     labels.append(y)

# Class index encoding
labels = []
for x in categories:
    y = CLASSES.index(x)
    labels.append(y)

In [7]:
np.random.seed(42)
train_texts, val_texts, train_labels, val_labels = train_test_split(articles, labels, test_size=0.1, random_state=42)
pickle.dump([train_texts, val_texts, train_labels, val_labels], open(DATA_PATH/'tokens'/'trnx_valx_trny_valy_ind_split.pkl', 'wb'))

### Tokenize

In [59]:
tok_train = Tokenizer(lang='xx').proc_all_mp(partition_by_cores(train_texts))
tok_val = Tokenizer(lang='xx').proc_all_mp(partition_by_cores(val_texts))

In [60]:
freq = Counter(p for o in tok_train for p in o)
print(len(tok_train))
freq.most_common(25)

43662


[(',', 657926),
 ('.', 559252),
 ('"', 217175),
 ('ja', 210514),
 ('on', 197759),
 ('et', 150766),
 ('ei', 106727),
 ('kui', 74991),
 ('ta', 66639),
 ('ka', 58212),
 ('oli', 51101),
 ('oma', 46727),
 ('-', 46020),
 ('ning', 45314),
 ('see', 45285),
 ('xbos', 43662),
 ('xfld', 43662),
 ('0', 42597),
 ('aga', 38936),
 ('t_up', 31812),
 ('mis', 31436),
 ('ma', 30478),
 ('siis', 29830),
 ('kes', 29218),
 ('tema', 28739)]

In [61]:
print(tok_val[5])

['xbos', 'vehklemisliidu', 'president', ',', 'riigikogu', 'liige', 'margus', 'hanson', 'tõdes', ',', 'et', 'naiskond', 'vehkles', 'kaunilt', 'kuni', 'finaalini', '.', '"', 'naised', 'olid', 'väga', 'tublid', '.', 'meil', 'on', 'noor', ',', 'perspektiivikas', 'ja', 'arenev', 'võistkond', ':', 'teise', 'kohaga', 'tuleb', 'igati', 'rahul', 'olla', ',', 'sest', 'ega', 'jõu', 'ja', 'võimu', 'vastu', 'ei', 'saa', '!', '"', 'hanson', 'lisas', ',', 'et', 'teda', 'rõõmustab', 'sten', 'priinitsa', 'individuaalturniiril', 'saadud', 'kaheksas', 'koht', ',', 'millega', 'mees', 'suurendab', 'ka', 't_up', 'eok', 'toetusraha', '.', '"', 'meie', 'vehklejad', 'on', 'tõestanud', ',', 'et', 'neid', 'saab', 'usaldada', '.', 'sportlased', 'seavad', 'kõrged', 'sihid', 'ja', 'on', 'võimelised', 'neid', 'täitma', ';', '"', 'kinnitas', 'ta', '.', 'ühtlasi', 'märkis', 'hanson', ',', 'et', 'suur', 'on', 'treener', 'igor', 'tšikinjovi', 'panus', '.', '"', 'ta', 'on', 'toonud', 'värsket', 'verd', 'ja', 'hingamist',

In [62]:
freq_val = Counter(p for o in tok_val for p in o)
print(len(tok_val))
freq_val.most_common(25)

4852


[(',', 72534),
 ('.', 62293),
 ('"', 23741),
 ('ja', 23599),
 ('on', 21847),
 ('et', 16600),
 ('ei', 11625),
 ('kui', 8402),
 ('ta', 7294),
 ('ka', 6594),
 ('oli', 5541),
 ('ning', 5219),
 ('oma', 5142),
 ('-', 5101),
 ('see', 4893),
 ('xbos', 4852),
 ('xfld', 4852),
 ('0', 4743),
 ('aga', 4345),
 ('mis', 3589),
 ('t_up', 3475),
 ('ma', 3323),
 ('tema', 3264),
 ('eesti', 3248),
 ('siis', 3235)]

In [63]:
np.save(DATA_PATH/'tokens/tok_train_pad.npy', tok_train)
np.save(DATA_PATH/'tokens/tok_val_pad.npy', tok_val)

In [64]:
itos = [o for o,c in freq.most_common(MAX_VOCAB) if c>min_freq]
itos.insert(0, '_pad_')
itos.insert(0, '_unk_')

In [65]:
stoi = collections.defaultdict(lambda:0, {v:k for k,v in enumerate(itos)})

In [97]:
train_lm = np.array([[stoi[o] for o in p] for p in tok_train])
val_lm = np.array([[stoi[o] for o in p] for p in tok_val])

In [117]:
# Pad and crop values
train_lm_pad = [x[:MAX_SIZE] if len(x) > MAX_SIZE else x + [0 for i in range(MAX_SIZE - len(x))] for x in train_lm]
val_lm_pad = [x[:MAX_SIZE] if len(x) > MAX_SIZE else x + [0 for i in range(MAX_SIZE - len(x))] for x in val_lm]

In [118]:
np.save(DATA_PATH/'tokens'/'trn_ids.npy', train_lm_pad) # Oversaved all as padded
np.save(DATA_PATH/'tokens'/'val_ids.npy', val_lm_pad)
pickle.dump(itos, open(DATA_PATH/'tokens'/'itos.pkl', 'wb'))

### Load tokenized data

In [6]:
train_texts, val_texts, train_labels, val_labels = pickle.load(open(DATA_PATH/'tokens'/'trnx_valx_trny_valy_ind_split.pkl', 'rb'))
train_lm = np.load(DATA_PATH/'tokens'/'trn_ids.npy')
val_lm = np.load(DATA_PATH/'tokens'/'val_ids.npy')
itos = pickle.load(open(DATA_PATH/'tokens'/'itos.pkl', 'rb'))

#### Display data

In [91]:
print(train_texts[0])

xbos Peaminister Taavi Rõivas jätab võimutüli tõttu ära visiidid Leedusse ja Rootsi, teda asendab väliskaubandus- ja ettevõtlusminister Anne Sulling.  Valitsuse pressiesindaja kinnitas pühapäeva pärastlõunal, et Rõivas ei sõida esmaspäeval visiidile Leetu ja Rootsi. Pressiesindaja teatel jäävad visiidid ära "seoses ametikohustustega Eestis". Reformierakonna esimees, peaminister Taavi Rõivas pidi esmaspäeval koos teiste Balti riikide valitsusjuhtidega osalema Leedus Klaipedas aset leidval LNG ujuvterminali saabumistseremoonial. Enne tseremooniat pidi aset leidma peaministrite ning Ameerika Ühendriikide esindajate ühine töölõuna. Pärastlõunal pidi Rõivas suunduma edasi Stockholmi, kus toimub Balti- ja Põhjamaade tippkohtumine. Rootsi, Soome, Norra, Islandi, Taani, Eesti, Läti ja Leedu peaministrite kohtumisel räägitakse majanduse olukorrast Euroopas, transatlantilistest suhetest ning Ukrainaga seotud arengutest. Pühapäeval kohtuvad Reformierakonna ja Sotsiaaldemokraatliku Erakonna esimeh

In [105]:
print(train_lm[0])

[17, 425, 524, 658, 2109, 0, 254, 63, 48013, 0, 5, 563, 2, 84, 28902, 0, 5, 53171, 1428, 26646, 3, 64, 755, 588, 438, 2029, 1368, 2, 7, 658, 8, 7061, 661, 4845, 17602, 5, 563, 3, 588, 704, 1070, 48013, 63, 4, 552, 0, 136, 4, 3, 813, 829, 2, 425, 524, 658, 388, 661, 79, 383, 555, 1197, 0, 3403, 5258, 0, 1815, 0, 21, 11591, 0, 0, 3, 105, 50278, 388, 1815, 3905, 31416, 15, 542, 1406, 7017, 3196, 0, 3, 1368, 388, 658, 0, 180, 3805, 2, 45, 638, 50279, 5, 9650, 11347, 3, 563, 2, 322, 2, 954, 2, 5409, 2, 2147, 2, 27, 2, 662, 5, 1109, 31416, 2357, 2275, 3893, 3806, 938, 2, 0, 5920, 15, 12824, 433, 43618, 3, 679, 10646, 813, 5, 7930, 871, 48014, 2, 7, 3819, 1726, 1064, 194, 0, 7214, 1899, 3, 18, 37, 7215, 19, 19, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0

In [10]:
print(train_labels[0])
print(CLASSES[train_labels[0]])
# print(CLASSES[train_labels[0].index(1)]) # for one hot

130
uudised/eesti


#### Dataloader

In [7]:
bs=128

class TokDataset(Dataset):
    def __init__(self, x, y):
        self.x = x; self.y = y
        self.len = len(self.x)
        self.x_data = torch.from_numpy(self.x); self.x_data = self.x_data.long()
        self.y_data = torch.from_numpy(self.y); self.y_data = self.y_data.long()
        print('x shape', self.x_data.shape)
        print('y shape', self.y_data.shape)
        
    def __getitem__(self, index):
        return self.x_data[index], self.y_data[index]
    
    def __len__(self):
        return self.len
    
ds = TokDataset(train_lm, np.asarray(train_labels))
ds_val = TokDataset(val_lm, np.asarray(val_labels))
dl = torch.utils.data.DataLoader(dataset=ds, batch_size=bs, shuffle=True, num_workers=0)
dl_val = torch.utils.data.DataLoader(dataset=ds_val, batch_size=bs, shuffle=True, num_workers=0)


x shape torch.Size([43662, 250])
y shape torch.Size([43662])
x shape torch.Size([4852, 250])
y shape torch.Size([4852])


In [11]:
test_values = iter(dl)
xs, ys = next(test_values)

### Fit function

In [25]:
def fit(epochs, model, train_dl, val_dl, crit, opt, verb=250):
    for ep in range(epochs):
        model.train()
        for i, data in tqdm(enumerate(train_dl)):
            x, y = data
            x = x.cuda(); y = y.cuda()

            y_h = model(x)
            loss = crit(y_h, y)
            
            # For accuracy
            total = 0
            correct = 0
            total += y.size(0)
            correct += (torch.argmax(y_h, 1) == y).sum().item()

            if i % verb == 0:
                print(f' Epoch: {ep} | b_loss: {loss.item():.{4}f}, b_acc: {100 * correct / total}')

            opt.zero_grad()
            loss.backward()
            opt.step()
        
        # Validate
        val_correct = 0
        model.eval()
        for i, data_val in enumerate(val_dl):
            x_val, y_val = data_val
            x_val = x_val.cuda(); y_val = y_val.cuda()
            
            # .eval() doesn't turn off gradient tracking
            with torch.no_grad():
                y_h_val = model(x_val)
                val_correct += (torch.argmax(y_h_val, 1) == y_val).sum().item()
        print(f'\nEPOCH {ep} - Val acc: {100 * val_correct / len(val_dl.dataset)}\n')
            

### Feed-forward NN

In [15]:
torch.__version__

'0.4.1'

In [116]:
class SimpleFNN(nn.Module):
    def __init__(self, input_size, vocab_size, num_outputs, num_l, neurons: List[int], e_size=100):
        super(SimpleFNN, self).__init__()
        self.e = nn.Embedding(vocab_size, e_size)
        self.input_l = nn.Linear(e_size * input_size, neurons[0])
        self.middle_l = nn.ModuleList()
        for i in range(num_l):
            self.middle_l.append(nn.Linear(neurons[i], neurons[i+1]))
        self.output_l = nn.Linear(neurons[-1], num_outputs)
        
        self.dropout = nn.Dropout(0.25)
        
    def forward(self, x):
        i_sz = x.shape[-1]
        x = F.relu(self.e(x))
        x = x.view(-1,  i_sz * x.shape[-1])
        x = F.relu(self.input_l(x))
        for l in self.middle_l:
            x = F.relu(self.dropout(l(x)))
        return self.output_l(x) # No softmax for crossentropy
        

In [117]:
fnn = SimpleFNN(MAX_SIZE, MAX_VOCAB, CLASS_COUNT, 4, [200, 300, 100, 50, 20]).cuda()
print(fnn)

SimpleFNN(
  (e): Embedding(60000, 100)
  (input_l): Linear(in_features=25000, out_features=200, bias=True)
  (middle_l): ModuleList(
    (0): Linear(in_features=200, out_features=300, bias=True)
    (1): Linear(in_features=300, out_features=100, bias=True)
    (2): Linear(in_features=100, out_features=50, bias=True)
    (3): Linear(in_features=50, out_features=20, bias=True)
  )
  (output_l): Linear(in_features=20, out_features=138, bias=True)
  (dropout): Dropout(p=0.25)
)


In [118]:
# Test pass through
fnn(xs.cuda()).shape

torch.Size([128, 138])

In [123]:
crit = torch.nn.CrossEntropyLoss()
opt = optim.Adam(fnn.parameters(), 1e-4)

In [124]:
fit(100, fnn, dl, dl_val, crit, opt)

0it [00:00, ?it/s] Epoch: 0 | b_loss: 0.4515, b_acc: 85.9375
246it [00:04, 55.67it/s] Epoch: 0 | b_loss: 0.1458, b_acc: 95.3125
342it [00:06, 55.55it/s]

EPOCH 0 - Val acc: 81.18301731244847

0it [00:00, ?it/s] Epoch: 1 | b_loss: 0.2479, b_acc: 93.75
246it [00:04, 56.20it/s] Epoch: 1 | b_loss: 0.3079, b_acc: 92.1875
342it [00:06, 55.98it/s]

EPOCH 1 - Val acc: 81.10057708161582

0it [00:00, ?it/s] Epoch: 2 | b_loss: 0.3286, b_acc: 89.0625
246it [00:04, 55.56it/s] Epoch: 2 | b_loss: 0.3491, b_acc: 90.625
342it [00:06, 55.46it/s]

EPOCH 2 - Val acc: 81.24484748557296

0it [00:00, ?it/s] Epoch: 3 | b_loss: 0.2795, b_acc: 91.40625
246it [00:04, 55.44it/s] Epoch: 3 | b_loss: 0.2152, b_acc: 91.40625
342it [00:06, 55.43it/s]

EPOCH 3 - Val acc: 81.38911788953008

0it [00:00, ?it/s] Epoch: 4 | b_loss: 0.2213, b_acc: 93.75
246it [00:04, 55.89it/s] Epoch: 4 | b_loss: 0.3165, b_acc: 91.40625
342it [00:06, 56.01it/s]

EPOCH 4 - Val acc: 81.30667765869744

0it [00:00, ?it/s] Epoch: 5 | b_loss: 0.25

0it [00:00, ?it/s] Epoch: 84 | b_loss: 0.1949, b_acc: 94.53125
246it [00:04, 55.84it/s] Epoch: 84 | b_loss: 0.3123, b_acc: 94.53125
342it [00:06, 55.91it/s]

EPOCH 84 - Val acc: 82.11046990931574

0it [00:00, ?it/s] Epoch: 85 | b_loss: 0.2599, b_acc: 90.625
246it [00:04, 55.98it/s] Epoch: 85 | b_loss: 0.2975, b_acc: 91.40625
342it [00:06, 56.00it/s]

EPOCH 85 - Val acc: 81.9249793899423

0it [00:00, ?it/s] Epoch: 86 | b_loss: 0.2347, b_acc: 89.84375
246it [00:04, 56.21it/s] Epoch: 86 | b_loss: 0.0674, b_acc: 96.875
342it [00:06, 56.12it/s]

EPOCH 86 - Val acc: 81.9249793899423

0it [00:00, ?it/s] Epoch: 87 | b_loss: 0.1208, b_acc: 97.65625
246it [00:04, 55.70it/s] Epoch: 87 | b_loss: 0.1463, b_acc: 96.09375
342it [00:06, 55.74it/s]

EPOCH 87 - Val acc: 81.88375927452597

0it [00:00, ?it/s] Epoch: 88 | b_loss: 0.2200, b_acc: 93.75
246it [00:04, 56.05it/s] Epoch: 88 | b_loss: 0.0901, b_acc: 98.4375
342it [00:06, 55.98it/s]

EPOCH 88 - Val acc: 81.760098928277

0it [00:00, ?it/s] Epoch: 8

### CNN model

In [13]:
# https://github.com/Shawn1993/cnn-text-classification-pytorch/blob/master/model.py
# https://arxiv.org/pdf/1408.5882.pdf

# Draft implementation
class SimpleCNN(nn.Module):
    def __init__(self, input_size, vocab_size, num_outputs, e_size=300, k_num=100, k_sizes=[3, 4, 5]):
        super(SimpleCNN, self).__init__()
        self.e = nn.Embedding(vocab_size, e_size)
        self.dropout_e = nn.Dropout(0.2)
        self.convs = nn.ModuleList([nn.Conv2d(1, k_num, (k, e_size)) for k in k_sizes])
        
        self.dropout = nn.Dropout(0.5)
        self.output_l = nn.Linear(len(k_sizes)*k_num, num_outputs)

    def forward(self, x):
        x = self.e(x)  # (N, W, D)
        x = self.dropout_e(x)
        x = x.unsqueeze(1)  # (N, Ci, W, D)
        
        x = [self.dropout(F.relu(conv(x)).squeeze(3)) for conv in self.convs]
        x = [F.max_pool1d(i, i.size(2)).squeeze(2) for i in x] 
        x = torch.cat(x, 1)

        x = self.dropout(x)  # (N, len(Ks)*Co)
        
        return self.output_l(x)

    
    def conv_and_pool(self, x, conv):
        x = F.relu(conv(x)).squeeze(3)  # (N, Co, W)
        x = F.max_pool1d(x, x.size(2)).squeeze(2)
        return x
        

In [14]:
cnn = SimpleCNN(MAX_SIZE, MAX_VOCAB, CLASS_COUNT).cuda()
print(cnn)

SimpleCNN(
  (e): Embedding(60000, 300)
  (dropout_e): Dropout(p=0.2)
  (convs): ModuleList(
    (0): Conv2d(1, 100, kernel_size=(3, 300), stride=(1, 1))
    (1): Conv2d(1, 100, kernel_size=(4, 300), stride=(1, 1))
    (2): Conv2d(1, 100, kernel_size=(5, 300), stride=(1, 1))
  )
  (dropout): Dropout(p=0.5)
  (output_l): Linear(in_features=300, out_features=138, bias=True)
)


In [15]:
cnn(xs.cuda()).shape

torch.Size([128, 138])

In [18]:
crit = torch.nn.CrossEntropyLoss()
opt = optim.Adam(cnn.parameters(), 1e-3)

In [51]:
fit(15, cnn, dl, dl_val, crit, opt)

0it [00:00, ?it/s] Epoch: 0 | b_loss: 0.1681, b_acc: 95.3125
250it [00:13, 18.65it/s] Epoch: 0 | b_loss: 0.2765, b_acc: 92.96875
342it [00:18, 18.65it/s]

EPOCH 0 - Val acc: 92.55976916735366

0it [00:00, ?it/s] Epoch: 1 | b_loss: 0.1518, b_acc: 95.3125
250it [00:13, 18.74it/s] Epoch: 1 | b_loss: 0.1871, b_acc: 92.96875
342it [00:18, 18.77it/s]

EPOCH 1 - Val acc: 92.31244847485573

0it [00:00, ?it/s] Epoch: 2 | b_loss: 0.3224, b_acc: 89.84375
250it [00:13, 18.05it/s] Epoch: 2 | b_loss: 0.1485, b_acc: 96.875
342it [00:18, 18.11it/s]

EPOCH 2 - Val acc: 92.35366859027205

0it [00:00, ?it/s] Epoch: 3 | b_loss: 0.1194, b_acc: 96.875
250it [00:13, 18.82it/s] Epoch: 3 | b_loss: 0.1504, b_acc: 94.53125
342it [00:18, 18.84it/s]

EPOCH 3 - Val acc: 92.45671887881286

0it [00:00, ?it/s] Epoch: 4 | b_loss: 0.3029, b_acc: 92.1875
250it [00:13, 18.43it/s] Epoch: 4 | b_loss: 0.1258, b_acc: 96.875
342it [00:18, 18.55it/s]

EPOCH 4 - Val acc: 92.35366859027205

0it [00:00, ?it/s] Epoch: 5 | b_loss: 0

In [34]:
torch.save(cnn, DATA_PATH/'models/cnn_es128_knum100_ksz345_90d8acc')

C:\Users\Ranet\Anaconda3\envs\fastai\lib\site-packages\torch\serialization.py:241: UserWarning: Couldn't retrieve source code for container of type SimpleCNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


### GRU Model

In [93]:
class GRU(nn.Module):
    def __init__(self, vocab_size, e_size, bs, n_layers, n_outputs, n_hidden=128):
        super().__init__()
        self.vocab_size, self.n_layers, self.n_hidden, self.n_outputs = vocab_size, n_layers, n_hidden, n_outputs
        self.e = nn.Embedding(vocab_size, e_size)
        self.rnn = nn.GRU(e_size, n_hidden, n_layers, dropout=0.5, batch_first=True)
        self.l_out = nn.Linear(n_hidden, n_outputs)
        
    def forward(self, cs):
        bs = cs.size(0)
        hidden = self.init_hidden(bs)
        outp, h = self.rnn(self.e(cs), hidden)
        outp = outp[:,-1,:] # all batches, last word, all output values
        fc_outp = self.l_out(outp)
        return fc_outp
        
            
    
    def init_hidden(self, bs):
        return torch.zeros(self.n_layers, bs, self.n_hidden).cuda()

In [94]:
gru = GRU(MAX_VOCAB, 300, bs, 2, CLASS_COUNT).cuda()
print(gru)

GRU(
  (e): Embedding(60000, 300)
  (rnn): GRU(300, 128, num_layers=2, batch_first=True, dropout=0.5)
  (l_out): Linear(in_features=128, out_features=138, bias=True)
)


In [95]:
test_val = gru(xs.cuda())
test_val.shape

torch.Size([128, 138])

In [96]:
crit = nn.CrossEntropyLoss()
opt = optim.Adam(gru.parameters(), 1e-3)

In [97]:
fit(15, gru, dl, dl_val, crit, opt)

0it [00:00, ?it/s] Epoch: 0 | b_loss: 4.9856, b_acc: 0.0
250it [00:18, 13.51it/s] Epoch: 0 | b_loss: 2.1697, b_acc: 45.3125
342it [00:25, 13.45it/s]

EPOCH 0 - Val acc: 57.04863973619126

0it [00:00, ?it/s] Epoch: 1 | b_loss: 1.7913, b_acc: 57.8125
250it [00:18, 13.49it/s] Epoch: 1 | b_loss: 0.9771, b_acc: 76.5625
342it [00:25, 13.52it/s]

EPOCH 1 - Val acc: 73.3305853256389

0it [00:00, ?it/s] Epoch: 2 | b_loss: 0.9018, b_acc: 78.90625
250it [00:18, 13.53it/s] Epoch: 2 | b_loss: 0.7622, b_acc: 78.90625
342it [00:25, 13.59it/s]

EPOCH 2 - Val acc: 80.87386644682606

0it [00:00, ?it/s] Epoch: 3 | b_loss: 0.5536, b_acc: 86.71875
250it [00:18, 13.64it/s] Epoch: 3 | b_loss: 0.5346, b_acc: 88.28125
342it [00:25, 13.62it/s]

EPOCH 3 - Val acc: 83.38829348722176

0it [00:00, ?it/s] Epoch: 4 | b_loss: 0.3831, b_acc: 90.625
250it [00:18, 13.27it/s] Epoch: 4 | b_loss: 0.5824, b_acc: 88.28125
342it [00:25, 13.27it/s]

EPOCH 4 - Val acc: 85.6760098928277

0it [00:00, ?it/s] Epoch: 5 | b_loss: 0.34

### LSTM Model

In [98]:
class LSTM(nn.Module):
    def __init__(self, vocab_size, e_size, bs, n_layers, n_outputs, n_hidden=128):
        super().__init__()
        self.vocab_size, self.n_layers, self.n_hidden, self.n_outputs = vocab_size, n_layers, n_hidden, n_outputs
        self.e = nn.Embedding(vocab_size, e_size)
        self.rnn = nn.LSTM(e_size, n_hidden, n_layers, dropout=0.5, batch_first=True)
        self.l_out = nn.Linear(n_hidden, n_outputs)
        self.init_hidden(bs)
        
    def forward(self, cs):
        bs = cs.size(0)
        if self.h[0].size(1) != bs: self.init_hidden(bs)
        outp, h = self.rnn(self.e(cs), self.h)
        outp = outp[:,-1,:] # all batches, last word, all output values
        #return F.log_softmax(self.l_out(outp), dim=-1).view(-1, self.n_outputs)
        return F.log_softmax(self.l_out(outp))
        
            
    
    def init_hidden(self, bs):
#         import pdb;pdb.set_trace()
        self.h = (torch.zeros(self.n_layers, bs, self.n_hidden).cuda(),
                  torch.zeros(self.n_layers, bs, self.n_hidden).cuda())

In [99]:
lstm = LSTM(MAX_VOCAB, 256, bs, 2, CLASS_COUNT).cuda()
print(lstm)

LSTM(
  (e): Embedding(60000, 256)
  (rnn): LSTM(256, 128, num_layers=2, batch_first=True, dropout=0.5)
  (l_out): Linear(in_features=128, out_features=138, bias=True)
)


In [100]:
test_val = lstm(xs.cuda())
test_val.shape

torch.Size([128, 138])

In [104]:
crit = F.nll_loss
opt = optim.Adam(lstm.parameters(), 1e-4)

In [105]:
fit(15, lstm, dl, dl_val, crit, opt)

0it [00:00, ?it/s] Epoch: 0 | b_loss: 0.0247, b_acc: 99.21875
250it [00:19, 13.06it/s] Epoch: 0 | b_loss: 0.0106, b_acc: 100.0
342it [00:25, 13.17it/s]

EPOCH 0 - Val acc: 89.98351195383347

0it [00:00, ?it/s] Epoch: 1 | b_loss: 0.0216, b_acc: 99.21875
250it [00:18, 13.23it/s] Epoch: 1 | b_loss: 0.0930, b_acc: 98.4375
342it [00:25, 13.21it/s]

EPOCH 1 - Val acc: 90.04534212695796

0it [00:00, ?it/s] Epoch: 2 | b_loss: 0.0042, b_acc: 100.0
250it [00:18, 13.30it/s] Epoch: 2 | b_loss: 0.0054, b_acc: 100.0
342it [00:25, 13.30it/s]

EPOCH 2 - Val acc: 90.14839241549876

0it [00:00, ?it/s] Epoch: 3 | b_loss: 0.0228, b_acc: 99.21875
250it [00:19, 12.87it/s] Epoch: 3 | b_loss: 0.0361, b_acc: 99.21875
342it [00:26, 12.94it/s]

EPOCH 3 - Val acc: 90.12778235779061

0it [00:00, ?it/s] Epoch: 4 | b_loss: 0.0350, b_acc: 99.21875
250it [00:19, 12.94it/s] Epoch: 4 | b_loss: 0.0464, b_acc: 98.4375
342it [00:26, 13.06it/s]

EPOCH 4 - Val acc: 90.14839241549876

0it [00:00, ?it/s] Epoch: 5 | b_loss: 0.0

# TODO:

##### Implementation
- Disable dropout during test time (proper eval)+
- Try a properly tuned model
    - cnn seems to get 92.47% acc, high variance (overfitting) on training data
        - e_size=300, bs=128, lr=1e-3 got same acc in 6 epochs opposed to ~40
    + might need a bigger batch size+
    + try more regularization
- Try learning rate finder
- Learning rate cosine annealing
- SGD with restarts
- Try to predict multiple labels

##### Analysis
+ Can also check how balanced the classes are+
    - Rework unbalanced classes or set class weights
+ Try a simple shallow learning model+
    - 94.5% accuracy, without fiddling with hyperparams and models much.




### Shallow SGD Classifier

In [46]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import CountVectorizer

text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-5, random_state=42,
                                           max_iter=25, tol=None)),])

In [47]:
text_clf.fit(train_texts, train_labels)  
predicted = text_clf.predict(val_texts)      

In [48]:
np.mean(predicted == np.asarray(val_labels))

0.9453833470733718